src/loan/ExitQueuePSM.sol

```diff
@@ -0,0 +1,303 @@
+        address owner; // owner of the CREDIT
+        uint64 feePercent; // fees percentage. 0.1e18 is 10%
+        uint64 timestamp;
+    }
+
+    /// @notice mapping of ticketId => ExitQueueTicket
+    /// this is where all the tickets data are stored
+    mapping(bytes32 => ExitQueueTicket) internal _tickets;
+
+    /// @notice ordered queue of ticket ids
+    /// front of the queue is the first ticket to be used when minting
+    /// back of the queue is the last ticket to be used when minting
+    /// @dev if the queue is empty, the contract act as a SimplePSM
+    DoubleEndedQueue.Bytes32Deque public queue;
+
+    /// @notice Event emitted upon entering the exit queue
+    /// @param owner Owner of the CREDIT
+    /// @param amount Amount of CREDIT being entered into the queue
+    /// @param feePercent Fee percentage for the redemption
+    /// @param timestamp Timestamp when the ticket is created
+    /// @param ticketId Unique identifier for the exit queue ticket
+    event EnterExitQueue(
+        address indexed owner,
+        uint256 amount,
+        uint64 feePercent,
+        uint256 timestamp,
+        bytes32 ticketId
+    );
+
+    /// @notice Constructor for ExitQueuePSM
+    /// @param _core Address of the core contract
+    /// @param _profitManager Address of the ProfitManager contract
+    /// @param _credit Address of the Credit token
+    /// @param _pegToken Address of the pegged token
+    /// @param _minAmount Minimum amount of CREDIT required to enter the exit queue
+    /// @param _minWithdrawDelay Minimum delay required before withdrawing from the exit queue
+    constructor(
+        address _core,
+        address _profitManager,
+        address _credit,
+        address _pegToken,
+        uint256 _minAmount,
+        uint256 _minWithdrawDelay
+    ) SimplePSM(_core, _profitManager, _credit, _pegToken) {
+        MIN_AMOUNT = _minAmount;
+        MIN_WITHDRAW_DELAY = _minWithdrawDelay;
+    }
+
+    /// @notice Retrieves the first ticket in the queue
+    /// @dev Utility function, to be used externally
+    /// @return ticket The first ticket in the exit queue
+    function getFirstTicket()
+        external
+        view
+        returns (ExitQueueTicket memory ticket)
+    {
+        bytes32 frontTicketId = queue.front();
+        return _tickets[frontTicketId];
+    }
+
+    /// @notice Retrieves the length of the queue
+    /// @dev Utility function, to be used externally
+    /// @return The current length of the queue
+    function getQueueLength() external view returns (uint256) {
+        return queue.length();
+    }
+
+    /// @notice Retrieves the total amount of credit in the queue and the total cost in peg token including the discount fees
+    /// @dev Utility function, to be used externally
+    /// @return totalCredit Total amount of CREDIT in the queue
+    /// @return totalCostPegToken Total cost in peg tokens for the CREDIT in the queue
+    function getTotalCreditInQueue()
```

Comment:
Hmmm it's done on purpose, the getTotalCreditInQueue is just to be used offchain. It's too costly to be used on chain

The getMintAmountOut will indeed not compute the real amountIn cost

I think the scenario, when using getMintAmountOut and the exit queue PSM will be:

1/ the user wants 1000 CREDIT, he uses the getMintAmountOut function to see how much USDC it will cost, let's say 1000 USDC
2/ here prepares the approval (1000 USDC to the psm) and start the mint
3/ if there are some tickets in the exit queue with fee > 0, then the user will have 1000 CREDIT but it will have cost him < 1000 USDC

Isn't that right?